lib/classes/class-db.php

```diff
@@ -0,0 +1,886 @@
+       * Get file data. If $size_names is set to true, all sizes will be included
+       * 
+       * @param array $meta
+       * @param int $attachment_id
+       * @param bool $with_sizes
+       * @return array
+       */
+      public function get_file($meta, $attachment_id, $with_sizes = false) {
+        if ( defined('WP_STATELESS_POSTMETA') && WP_STATELESS_POSTMETA ) {
+          return get_post_meta($attachment_id, 'sm_cloud', true);
+        }
+
+        // Get values from the cache
+        $cache_key = $this->_get_file_cache_key($attachment_id);
+
+        $meta = wp_cache_get($cache_key, $this->cache_group, false, $found);
+
+        if ( $found ) {
+          return $meta;
+        }
+
+        // Get values from the DB
+        $fields = $this->_map_fields([], $this->file_mapping);
+
+        $sql = "SELECT $fields FROM $this->files WHERE post_id = %d";
+
+        $meta = $this->wpdb->get_row( $this->wpdb->prepare($sql, $attachment_id), ARRAY_A );
+
+        wp_cache_set($cache_key, $meta, $this->cache_group);
+
+        if ( empty($meta) ) {
+          return get_post_meta($attachment_id, 'sm_cloud', true);
+        }
+
+        // Get file size meta data
+        if ( $with_sizes ) {
+          $meta['sizes'] = apply_filters('wp_stateless_get_file_sizes', [], $attachment_id);
+        }
+
+        return $meta;
+      }
+
+      /**
+       * Get file sizes data
+       * 
+       * @param array $sizes
+       * @param int $attachment_id
+       * @return array
+       */
+      public function get_file_sizes($sizes, $attachment_id) {
+        if ( defined('WP_STATELESS_POSTMETA') && WP_STATELESS_POSTMETA ) {
+          $meta = get_post_meta($attachment_id, 'sm_cloud', true);
+          return isset($meta['sizes']) ? $meta['sizes'] : [];
+        }
+
+        // Get values from the cache
+        $cache_key = $this->_get_file_sizes_cache_key($attachment_id);
+
+        $sizes = wp_cache_get($cache_key, $this->cache_group, false, $found);
+
+        if ( $found ) {
+          return $sizes;
+        }
+
+        // Get values from the DB
+        $fields = $this->_map_fields([], $this->file_sizes_mapping);
+
+        $sql = "SELECT $fields FROM $this->file_sizes WHERE post_id = %d";
+        $sql = $this->wpdb->prepare($sql, $attachment_id);
+
+        $result = $this->wpdb->get_results( $this->wpdb->prepare($sql, $attachment_id), ARRAY_A );
+        $sizes = [];
+
+        if ( !empty($result) ) {
+          foreach ($result as $size) {
+            $size_name = $size['size_name'];
+            unset($size['size_name']);
+            $sizes[$size_name] = $size;
+          }
+        }
+
+        wp_cache_set($cache_key, $sizes, $this->cache_group);
+
+        return $sizes;
+      }
+
+      /**
+       * Get file meta
+       * 
+       * @param int $post_id
+       * @param string $key
+       * @param mixed $default
+       * @return mixed
+       */
+      public function get_file_meta_value($post_id, $key, $default) {
```

Comment:
@balexey88 let's do the field `$default` optional

Reply:
And change it to `filter`.
So we will have 4 filters:
- wp_stateless_get_file
- wp_stateless_get_file_sizes
- wp_stateless_get_file_meta
- wp_stateless_get_file_meta_value

Reply:
Done https://github.com/udx/wp-stateless/commit/5266f8816a4e216aa663f1251d1e6b73de99e218